app/models/git_repository.rb

```diff
@@ -0,0 +1,104 @@
+class GitRepository < ApplicationRecord
+  include AuthenticationMixin
+
+  validates :url, :format => URI::regexp(%w(http https)), :allow_nil => false
+
+  default_value_for :verify_ssl, OpenSSL::SSL::VERIFY_PEER
+
+  has_many :git_branches, :dependent => :destroy
+  has_many :git_tags, :dependent => :destroy
+
+  VALID_KEYS = %w(commit_sha commit_message commit_time).freeze
+
+  def refresh
+    @repo = Dir.exist?(dirname) ? update_repo : init_repo
+    refresh_branches
+    refresh_tags
+    self.last_refresh_on = Time.now.utc
+    save!
+  end
+
+  def branch_info(name)
+    refresh unless @repo
+    branch = git_branches.detect { |item| item.name == name }
+    raise "Branch #{name} not found" unless branch
+    branch.attributes.slice(*VALID_KEYS)
+  end
+
+  def tag_info(name)
+    refresh unless @repo
+    tag = git_tags.detect { |item| item.name == name }
+    raise "GIT Tag #{name} not found" unless tag
+    tag.attributes.slice(*VALID_KEYS)
+  end
+
+  def dirname
+    @directory ||= directory_name
+  end
```

Comment:
I don't like the fact that this creates 3 names for the same thing (dirname, directory, directory_name)...I think you can just expose the directory_name method directly and use that where it's needed in the code...no need for caching, and no need for the shortened "dirname" method.